File: generated-code/kotlin/codecraft/TcpReadWrite/src/main/kotlin/model/AttackProperties.kt
package model

import util.StreamUtil

/**
 * Entity's attack properties
 */
class AttackProperties {
    /**
     * Maximum attack range
     */
    var attackRange: Int = 0
    /**
     * Damage dealt in one tick
     */
    var damage: Int = 0
    /**
     * If true, dealing damage will collect resource from target
     */
    var collectResource: Boolean = false

    constructor(attackRange: Int, damage: Int, collectResource: Boolean) {
        this.attackRange = attackRange
        this.damage = damage
        this.collectResource = collectResource
    }

    /**
     * Write AttackProperties to output stream
     */
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, attackRange)
        StreamUtil.writeInt(stream, damage)
        StreamUtil.writeBoolean(stream, collectResource)
    }

    /**
     * Get string representation of AttackProperties
     */
    override fun toString(): String {
        var stringBuilder = StringBuilder("AttackProperties { ")
        stringBuilder.append("attackRange: ")
        stringBuilder.append(attackRange)
        stringBuilder.append(", ")
        stringBuilder.append("damage: ")
        stringBuilder.append(damage)
        stringBuilder.append(", ")
        stringBuilder.append("collectResource: ")
        stringBuilder.append(collectResource)
        stringBuilder.append(" }")
        return stringBuilder.toString()
    }

    companion object {
        /**
         * Read AttackProperties from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): AttackProperties {
            var attackRange: Int
            attackRange = StreamUtil.readInt(stream)
            var damage: Int
            damage = StreamUtil.readInt(stream)
            var collectResource: Boolean
            collectResource = StreamUtil.readBoolean(stream)
            return AttackProperties(attackRange, damage, collectResource)
        }
    }
}

File: generated-code/go/codecraft/TcpReadWrite/model/attack_properties.go
package model

import "fmt"
import "io"
import . "trans_gen_test/stream"

// Entity's attack properties
type AttackProperties struct {
    // Maximum attack range
    AttackRange int32
    // Damage dealt in one tick
    Damage int32
    // If true, dealing damage will collect resource from target
    CollectResource bool
}

func NewAttackProperties(attackRange int32, damage int32, collectResource bool) AttackProperties {
    return AttackProperties {
        AttackRange: attackRange,
        Damage: damage,
        CollectResource: collectResource,
    }
}

// Read AttackProperties from reader
func ReadAttackProperties(reader io.Reader) AttackProperties {
    var attackRange int32
    attackRange = ReadInt32(reader)
    var damage int32
    damage = ReadInt32(reader)
    var collectResource bool
    collectResource = ReadBool(reader)
    return AttackProperties {
        AttackRange: attackRange,
        Damage: damage,
        CollectResource: collectResource,
    }
}

// Write AttackProperties to writer
func (attackProperties AttackProperties) Write(writer io.Writer) {
    attackRange := attackProperties.AttackRange
    WriteInt32(writer, attackRange)
    damage := attackProperties.Damage
    WriteInt32(writer, damage)
    collectResource := attackProperties.CollectResource
    WriteBool(writer, collectResource)
}

// Get string representation of AttackProperties
func (attackProperties AttackProperties) String() string {
    stringResult := "{ "
    stringResult += "AttackRange: "
    attackRange := attackProperties.AttackRange
    stringResult += fmt.Sprint(attackRange)
    stringResult += ", "
    stringResult += "Damage: "
    damage := attackProperties.Damage
    stringResult += fmt.Sprint(damage)
    stringResult += ", "
    stringResult += "CollectResource: "
    collectResource := attackProperties.CollectResource
    stringResult += fmt.Sprint(collectResource)
    stringResult += " }"
    return stringResult
}